### PR TITLE
[mlir] Drop outermost dims in slice rank reduction inference

### PR DESCRIPTION
**Pull request by Max191**
The `getDroppedDims` utility function does not follow the convention of dropping outermost unit dimensions first when inferring a rank reduction mask for a slice. This PR updates the implementation to match this convention.

### CODE REVIEW

**Comment by matthias-springer** [comment]
> convention of dropping outermost unit dimensions first when inferring a rank reduction mask for a slice

Where is this convention coming from? I always thought it is unspecified which dimension is dropped.

**Comment by Max191** [comment]
> > convention of dropping outermost unit dimensions first when inferring a rank reduction mask for a slice
> 
> Where is this convention coming from? I always thought it is unspecified which dimension is dropped.

This is not very well documented throughout the codebase, but it is documented here: https://mlir.llvm.org/docs/Dialects/TensorOps/#verification-vs-inference-in-the-rank-reduced-case. I have felt this is sort of an undocumented convention, but it should probably be better documented and enforced.

Also, extract/insert ops should probably have some additional arguments to specify which dimensions are being inserted along, but that is out of the scope of this PR. Probably a good discussion for another time, though.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -131,40 +131,40 @@ bool tensor::isSameTypeWithoutEncoding(Type tp1, Type tp2) {
 static llvm::SmallBitVector getDroppedDims(ArrayRef<int64_t> reducedShape,
                                            ArrayRef<OpFoldResult> mixedSizes) {
   llvm::SmallBitVector droppedDims(mixedSizes.size());
-  int64_t shapePos = 0;
+  int64_t shapePos = reducedShape.size() - 1;
 
-  for (const auto &size : enumerate(mixedSizes)) {
+  for (const auto &size : enumerate(llvm::reverse(mixedSizes))) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:136]: I personally find `enumerate(reverse(...))` hard to keep track off in my head.

File: mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir [hunk 2]
```diff
@@ -282,23 +282,6 @@ func.func @insert_slice_of_insert_slice(%t: tensor<f32>, %r0: tensor<1x1xf32>, %
 
 // -----
 
-// CHECK-LABEL: func @insert_slice_of_insert_slice(
```
Comment (MaheshRavishankar) [inline mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir:285]: Can you say why this was removed... The logic here is a bit tricky, This itself seems unrelated.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
The logic looks right to me, but want more clarity on the removed test.

File: mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir (hunk 2 above)
Comment (Max191) [reply mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir:285]: This test is failing with the new changes, but I think the test itself is ambiguous and doesn't make much sense. This is testing that an out of bounds offset into a single element tensor gets composed specifically with the out of bounds offset into the unit dimension of another tensor when the two inserts are combined.

(1) Based on the conventional way of inferring rank reduced dimensions, the offset should be composed with the inner offset into the `14` dim, so the test is incorrect.
(2) I would have simply updated the test, but the test doesn't make much sense to me anyway. It is composing 2 out of bounds inserts with each other to make an even more out of bounds insert. The behavior of both of these inserts is undefined, so composing them into a new, different undefined insert seems like a strange test to me.

If you see a reason to keep the test, then I can add it back and fix it, but otherwise I figured it made more sense to just remove it.

Comment (Max191) [reply mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir:285]: To clarify, it fails because the first insert is composed along the `14` dimension now, instead of the `1` dimension as it was before.

Comment (MaheshRavishankar) [reply mlir/test/Dialect/Tensor/fold-tensor-subset-ops.mlir:285]: Thanks for the explanation! The composition is load bearing. I think the previous test was wrong. IIUC the semantics of rank-reducing extract and insert is that the outer dimensions are dropped. Maybe fix the offsets and keep the test. It's good to test the composition

**Review by MaheshRavishankar — APPROVED** [review]
